File: java/leetcode/BinaryTreeFactors.java
//Leetcode problem 823 Binary Tree With Factors
//Solution written by Xuqiang Fang on 1 June, 2018 
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
class Solution{
    public int numFactoredBinaryTrees(int[] A) {
        long mod = (int)Math.pow(10, 9) + 7;
        int n = A.length;
        Arrays.sort(A);
        long[] dp = new long[n];
        Map<Integer, Integer> map = new HashMap<>();
        Arrays.fill(dp, 1);
        map.put(A[0], 0);
        int ans = 1;
        for(int i=1; i<n; ++i){
            for(int j=0; j<i; ++j){
                if(A[i]%A[j] == 0 && map.containsKey(A[i]/A[j])){
                    dp[i] += dp[j] * dp[map.get(A[i]/A[j])];
                    dp[i] %= mod;
                }
            }
            map.put(A[i], i);
            ans += dp[i];
            ans %= mod;
        }
        return ans;
    }   
}

public class BinaryTreeFactors{
	public static void main(String[] args){
		Solution s = new Solution();
        int[] nums = {2,3,4,5,8,10,12,16,20,24,28,30,32,36,40,50};
        System.out.println(s.numFactoredBinaryTrees(nums));
	}
}

File: py/leetcode/BinaryTreeFactors.py
'''
Leetcode problem No 823 Binary Tree With Factors
Solution written by Xuqiang Fang on 1 June, 2018
'''

class Solution(object):
    def numFactoredBinaryTrees(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        mo = pow(10,9)+7
        n = len(A)
        A.sort()
        m = {A[0]:0}
        dp = [1 for i in range(n)]
        ans = 1
        for i in range(1,n):
            for j in range(i):
                if A[i]/A[j] in m:
                    dp[i] += dp[j] * dp[m[A[i]/A[j]]] 
                    dp[i] %= mo
            m[A[i]] = i
            ans += dp[i]
            ans %= mo
        return ans

def main():
    s = Solution()
    A = [2,3,4,5,8,10,12,16,20,24,28,30,32,36,40,50]
    print(s.numFactoredBinaryTrees(A))
main()
